File: src/kattis/jazz/lab6/countingtriangles.py
from itertools import combinations

class Line:
    """
    A Line is a pair of (x,y) coordinates that correlate to the ends of the line
    """
    def __init__(self, str_array):
        self.x1 = float(str_array[0])
        self.y1 = float(str_array[1])
        self.x2 = float(str_array[2])
        self.y2 = float(str_array[3])

    def __str__(self):
        return f"({self.x1}, {self.y1}) - ({self.x2}, {self.y2})"

    @staticmethod
    def ccw(val1, val2, val3, val4, val5, val6):
        """
        Code cited here: https://bryceboe.com/2006/10/23/line-segment-intersection-algorithm/


        :param val1: The first x coordinate
        :param val2: The first y coordinate
        :param val3: The second x coordinate
        :param val4: The second y coordinate
        :param val5: The third x coordinate
        :param val6: The third y coordinate
        :return: True to show that a point intersects a line
        """
        return (val6 - val2) * (val3 - val1) > (val4 - val2) * (val5 - val1)


    def does_intersect(self, other_line):
        """

        :param other_line: The other line to check an intersection between
        :return: True if there's an intersection between the 2 lines
        """
        return Line.ccw(self.x1, self.y1, other_line.x1, other_line.y1, other_line.x2, other_line.y2) != Line.ccw(self.x2, self.y2, other_line.x1, other_line.y1, other_line.x2, other_line.y2) and \
    Line.ccw(self.x1, self.y1, self.x2, self.y2, other_line.x1, other_line.y1) != Line.ccw(self.x1, self.y1, self.x2, self.y2, other_line.x2, other_line.y2)


def are_all_connected(line1, line2, line3):
    """
    Determines whether the lines given are connected to form a triangle by seeing if they intersect with
    each other

    :param line1: The first line
    :param line2: The second line
    :param line3: The third
    :return: True if all the lines are connected to form a triangle
    """
    return line1.does_intersect(line2) and line1.does_intersect(line3) and line2.does_intersect(line3)


def analyze(combo_set):
    count = 0

    for combo in combo_set:
        if are_all_connected(*combo):
            count += 1
    print(count)


def main():
    line_count = int(input())
    all_line_groups = []
    group = []
    # Gathering all the inputs and turn them into lines objects
    while line_count != 0:
        for _ in range(line_count):
            group.append(Line(input().split()))
        all_line_groups.append(group)
        group = []
        line_count = int(input())

    # Creating the combinations with an r = 3 for a set of elements
    # Then, analyzing the combination set to find the combinations that form a triangle
    for line_group in all_line_groups:
        analyze(combinations(line_group, 3))


if __name__ == '__main__':
    main()

File: src/kattis/jazz/lab6/countingtriangles.java
package kattis.jazz.lab6;

import java.io.BufferedOutputStream;
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;

public class countingtriangles {
    private static final byte[] NEWLINE = "\n".getBytes();
    private static final BufferedReader IN = new BufferedReader(new InputStreamReader(System.in));
    private static final OutputStream OUT = new BufferedOutputStream(System.out);

    public static void main(String[] args) throws IOException {
        int lineCount = Integer.parseInt(IN.readLine());
        List<Line[]> lineSetsList = new ArrayList<>();
        Line[] lines = new Line[lineCount];

        //Gathering all inputs
        while (lineCount != 0) {
            for (int i = 0; i < lineCount; i++) {
                lines[i] = new Line(IN.readLine().split(" "));
            }
            lineCount = Integer.parseInt(IN.readLine());
            lineSetsList.add(lines);
            lines = new Line[lineCount];
        }


        //Iterating through each set of lines
        for (var lineSet : lineSetsList) {
            int counter = 0;
            //Calculating the indexes for combinations
            List<int[]> combinationIndexes = generateCombinationIndexes(lineSet.length);

            //Iterating all combinations to see if those lines intersect with each other to form a triangle
            for (int[] combo : combinationIndexes) {
                if (lineSet[combo[0]].doesIntersect(lineSet[combo[1]]) &&
                        lineSet[combo[0]].doesIntersect(lineSet[combo[2]]) &&
                        lineSet[combo[2]].doesIntersect(lineSet[combo[1]])) {
                    counter++;
                }
            }
            OUT.write(String.valueOf(counter).getBytes());
            OUT.write(NEWLINE);
        }

        OUT.flush();
    }

    /**
     * Code cited here: <a href="https://www.baeldung.com/java-combinations-algorithm">...</a>
     * Generates the index combinations for a set unique elements
     * The Combination equation uses 2 parameters, but since r is always 3, n is the only variable
     * that needs to be specified
     *
     * @param n The length of array used to produce combinations
     * @return The List of arrays, all with a length of 3
     */
    private static List<int[]> generateCombinationIndexes(int n) {
        List<int[]> combinations = new ArrayList<>();
        int[] combination = new int[3];

        // initialize with the lowest lexicographic combination
        for (int i = 0; i < 3; i++) {
            combination[i] = i;
        }

        while (combination[2] < n) {
            combinations.add(combination.clone());

            // generate next combination in lexicographic order
            int t = 2;
            while (t != 0 && combination[t] == n - 3 + t) {
                t--;
            }
            combination[t]++;
            for (int i = t + 1; i < 3; i++) {
                combination[i] = combination[i - 1] + 1;
            }
        }

        return combinations;
    }

    /**
     * A Line is a pair of (x,y) coordinates that correlate to the ends of the line
     */
    private static class Line {
        private final double x1, y1, x2, y2;

        public Line(String[] array) {
            x1 = Double.parseDouble(array[0]);
            y1 = Double.parseDouble(array[1]);
            x2 = Double.parseDouble(array[2]);
            y2 = Double.parseDouble(array[3]);
        }

        /**
         * Code cited here: <a href="https://bryceboe.com/2006/10/23/line-segment-intersection-algorithm/">...</a>
         *
         * @param value1 The first x coordinate
         * @param value2 The first y coordinate
         * @param value3 The second x coordinate
         * @param value4 The second y coordinate
         * @param value5 The third x coordinate
         * @param value6 The third y coordinate
         * @return True to show that a point intersects a line
         */
        private static boolean ccw(double value1, double value2, double value3,
                                   double value4, double value5, double value6) {
            return (value6 - value2) * (value3 - value1) > (value4 - value2) * (value5 - value1);
        }

        /**
         * Code cited here: <a href="https://bryceboe.com/2006/10/23/line-segment-intersection-algorithm/">...</a>
         *
         * @param other The other line to check an intersection between
         * @return True if there's an intersection between the 2 lines
         */
        public boolean doesIntersect(Line other) {
            return ccw(this.x1, this.y1, other.x1, other.y1, other.x2, other.y2) !=
                    ccw(this.x2, this.y2, other.x1, other.y1, other.x2, other.y2) &&
                    ccw(this.x1, this.y1, this.x2, this.y2, other.x1, other.y1) !=
                            ccw(this.x1, this.y1, this.x2, this.y2, other.x2, other.y2);
        }
    }
}
